disco_aws_automation/disco_elasticsearch.py

```diff
@@ -0,0 +1,418 @@
+"""
+Manage AWS ElasticSearch
+"""
+import logging
+import time
+import json
+from ConfigParser import NoOptionError
+
+import boto3
+
+from boto3.exceptions import Boto3Error
+from botocore.exceptions import BotoCoreError
+from . import read_config
+from .disco_route53 import DiscoRoute53
+from .resource_helper import throttled_call
+from .disco_aws_util import is_truthy
+from .disco_constants import DEFAULT_CONFIG_SECTION
+
+CONFIG_FILE = "disco_elasticsearch.ini"
+
+
+class DiscoElasticsearch(object):
+    """
+    A simple class to manage ElasticSearch
+    """
+
+    def __init__(self, environment_name, config_aws=None, config_es=None,
+                 route53=None):
+        self.config_aws = config_aws or read_config()
+        self.config_es = config_es or read_config(CONFIG_FILE)
+        self.route53 = route53 or DiscoRoute53()
+
+        if environment_name:
+            self.environment_name = environment_name.lower()
+        else:
+            self.environment_name = self.config_aws.get("disco_aws", "default_environment")
+
+        self._conn = None  # Lazily initialized
+        self._session = None  # Lazily initialized
+        self._account_id = None  # Lazily initialized
+        self._region = None  # Lazily initialized
+        self._zone = None  # Lazily initialized
+
+    @property
+    def conn(self):
+        """The boto3 elasticsearch connection object"""
+        if not self._conn:
+            self._conn = boto3.client('es')
+        return self._conn
+
+    @property
+    def session(self):
+        """Boto3 session"""
+        if not self._session:
+            self._session = boto3.session.Session()
+        return self._session
+
+    @property
+    def account_id(self):
+        """Account id of the current IAM user"""
+        if not self._account_id:
+            self._account_id = boto3.resource('iam').CurrentUser().arn.split(':')[4]
+        return self._account_id
+
+    @property
+    def region(self):
+        """Current region used by Boto"""
+        if not self._region:
+            # Doing this requires boto3>=1.2.4
+            # Could use undocumented and unsupported workaround for earlier versions:
+            # session._session.get_config_variable('region')
+            self._region = self.session.region_name
+        return self._region
+
+    @property
+    def zone(self):
+        """The current Route 53 zone"""
+        if not self._zone:
+            self._zone = self.get_aws_option('domain_name')
+        return self._zone
+
+    def get_domain_name(self, elasticsearch_name):
+        """
+        Get the name of the ElasticSearch domain.
+        Follows the format 'es-{elasticsearch_name}-{environment_name}'
+        """
+        return "es-{}-{}".format(elasticsearch_name, self.environment_name)
```

Comment:
Doing `{elasticsearch_name}-{environment_name}` was meant to be inline with the ELB naming schema, which has a similar pattern with the environment name at the end.

The point of the `es` prefix was to provide namespacing for elasticsearch domains in Route 53, because there's currently checks in place to assume someone doesn't do something strange and use the name of a hostclass as the name of elasticsearch domain.